packages/@aws-cdk/aws-apigatewayv2/lib/http/authorizer.ts

```diff
@@ -0,0 +1,119 @@
+import { Resource } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { CfnAuthorizer } from '../apigatewayv2.generated';
+
+import { IAuthorizer } from '../common';
+import { IHttpApi } from './api';
+
+/**
+ * Supported Authorizer types
+ */
+export enum HttpAuthorizerType {
+  /** JSON Web Tokens */
+  JWT = 'JWT',
+
+  /** Lambda Authorizer */
+  LAMBDA = 'REQUEST',
+}
+
+/**
+ * Properties to initialize an instance of `HttpAuthorizer`.
+ */
+export interface HttpAuthorizerProps {
+  /**
+   * Name of the authorizer
+   *
+   * @default - id of the HttpAuthorizer construct.
+   */
+  readonly authorizerName?: string
+
+  /**
+   * HTTP Api to attach the authorizer to
+   */
+  readonly httpApi: IHttpApi
+}
+
+/**
+ * Describes an instance of `IHttpAuthorizer`
+ */
+export interface IHttpAuthorizer extends IAuthorizer {
+  /**
+   * Type of authorizer
+   * @attribute
+   */
+  readonly authorizerType: HttpAuthorizerType;
+}
+
+/**
+ * Specifies the configuration of a JWT authorizer
+ */
+export interface JwtConfiguration {
+  /**
+   * A list of the intended recipients of the JWT
+   */
+  readonly audience: string[]
+
+  /**
+   * The base domain of the identity provider that issues JSON Web Tokens.
+   */
+  readonly issuer: string;
+}
+
+/**
+ * Properties to initialize an instance of `HttpAuthorizer`.
+ */
+export interface HttpJwtAuthorizerProps extends HttpAuthorizerProps {
+  /**
+   * The identity source for which authorization is requested.
+   *
+   * @default ['$request.header.Authorization']
+   */
+  readonly identitySource?: string[],
+
+  /**
+   * Configuration of a JWT authorizer.
+   * Required for the JWT authorizer type. Supported only for HTTP APIs.
+   *
+   * @default - No configuration
+   */
+  readonly jwtConfiguration: JwtConfiguration
+}
+
+/**
+ * Create a new JwtAuthorizer for Http APIs
+ * @resource AWS::ApiGatewayV2::Authorizer
+ */
+export class HttpJwtAuthorizer extends Resource implements IHttpAuthorizer {
+
+  /**
+   * Import an existing HTTP Authorizer into this CDK app.
+   */
+  public static fromAuthorizerId(scope: Construct, id: string, authorizerId: string): IHttpAuthorizer {
+    class Import extends Resource implements IHttpAuthorizer {
+      public readonly authorizerId = authorizerId;
+      public readonly authorizerType = HttpAuthorizerType.JWT;
+    }
+    return new Import(scope, id);
+  }
```

Comment:
This needs to sit either at the `HttpAuthorizer` level or `Authorizer` class. 

Neither exist at the moment, so I suggest pulling this out of this PR and adding it later as a separate PR. This will help us keep the number of changes contained and easy to approve.

Reply:
Okay, I see that'd be the abstract class pattern. 

Can we pause on this PR then and implement that first? It'd be great to have `fromAuthorizerId` in from day 1 of the release of this.

Here's what I'm thinking of how we go forward:

They're 3 authorizers:
- IAM
- Jwt
- Lambda

All should support `fromAuthorizerId` so we probably can make a base class called `AuthorizerBase` (this is what the separate PR would be for )

The "Http" prefix, I think might be a bit misleading as `AWS::ApiGatewayV2::Authorizer` is for both Http Apis and Websocket Apis.

The end product would be 3 authorizer classes:
- IamAuthorizer
- JwtAuthorizer (only working on this one currently)
- LambdaAuthorizer

Let me know if you agree with this general direction.

Reply:
We don't need 3 imports, just one. The type of an imported authorizer is not useful to the CDK as it's not configuring the authorizer.

There's a general re-organization of the constructs that are needed here that will make all of this cleaner.

What we need is a construct `class HttpAuthorizer extends Resource implements IAuthorizer` This class is what will have the import statement.
Next, you need to introduce an interface `IRouteAuthorizer` which will then be implemented by `HttpJwtAuthorizer` (and the other two types - IAM and Lambda - in the future) that will have a `bind()` operation. The implementation of the `bind()` will include creating a new `HttpAuthorizer`.
The `HttpRouteProps` will take a property of type `IRouteAuthorizer`.

Take a look at how `Integration` is implemented in this package. This is the same pattern you need for Authorizer.
https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-apigatewayv2/lib/http/integration.ts

Further, following on the same pattern as integrations - create a separate package `aws-apigatewayv2-authorizers` that will hold the individual authorizers and create a `UserPoolAuthorizer` that will take care of automatically configuring both the `audience` and `issuer`.

Reply:
Let me know if you need my help in starting this off a little.

Reply:
Okay cool I think I understand the changes needed. Will try to make these in a checkpoint fashion. Will leave the `aws-apigatewayv2-authorizers` package part for last or in another PR

Reply:
> The implementation of the bind() will include creating a new HttpAuthorizer.

Also, if I'm not mistaken, this seems to be going the path of the Route creating the authorizer for you (this is what the integration example you linked does). However, the same Authorizer should be able to be used across multiple routes (unlike integration), also this why `bind()` doesn't make sense to me when it comes to authorizers. 

Ideally, a user should be able to create an authorizer (standalone) and attach it to any route they decide.

Reply:
> Ideally, a user should be able to create an authorizer (standalone) and attach it to any route they decide.

Yes, this would be possible. Each route would call `bind()` on the authorizer and that would be their mechanism of binding. 
Like I mentioend, take a look at integration. It's pretty much the same pattern.

Reply:
Could you outline how the top-level usage looks like between Authorizer and Route?

This is how it is currently:

```ts
const authorizer = new HttpJwtAuthorizer(stack, 'HttpAuthorizer', {
  httpApi,
  jwtConfiguration: {
    audience: ['cognito-pool'],
    issuer: 'http://congnito.aws',
  },
});

new HttpRoute(stack, 'HttpRoute', {
  httpApi,
  integration: new DummyIntegration(),
  routeKey: HttpRouteKey.with('/books', HttpMethod.GET),
  authorizer,
  authorizationScopes: ['books:read'],
});
```

Maybe if you show how your implementation would work at a high level then I can work backwards.

> Like I mentioend, take a look at integration. It's pretty much the same pattern.

I've read over this code so many times but authorizer and integration seem like completely two different things.

Say I have 10 routes in my API, and 1 authorizer:

**Integration**
10 integrations, each one created and attached at the time of Route creation. `bind` makes sense here as we create the integration inside the route https://github.com/aws/aws-cdk/blob/ec61c3c054b1c04da60e9e0ae6e5b77bacc4f030/packages/%40aws-cdk/aws-apigatewayv2/lib/http/route.ts#L128

**Authorizer**
1 authorizer, created outside of and before Route creation, attached at Route creation. What exactly would be the `bind` do as no creation happens inside the Route?

<sub>Really apologize if I'm not getting you.</sub>

Reply:
> Really apologize if I'm not getting you.

No problem @iRoachie.

Let me get back to you on this next week. I'll see if I can make some changes directly and get you on the right track.